libs/base/Control/Function.idr

```diff
@@ -1,10 +1,21 @@
 module Control.Function
 
+||| An injective function maps distinct elements to distinct elements.
 public export
-interface Injective (f : a -> b) where
+interface Injective (f : a -> b) | f where
   constructor MkInjective
   injective : {x, y : a} -> f x = f y -> x = y
 
+----------------------------------------
+
+||| The composition of injective functions is injective.
+public export
+{f : a -> b} -> {g : b -> c} ->
+  (Injective f, Injective g) => Injective (g . f) where
+    injective prf = injective $ injective prf
+
+||| If (g . f) is injective, so is f.
 public export
-inj : (0 f : a -> b) -> {auto 0 _ : Injective f} -> {0 x, y : a} -> (0 _ : f x = f y) -> x = y
-inj _ eq = irrelevantEq (injective eq)
+{f : a -> b} -> {g : b -> c} ->
+  Injective (g . f) => Injective f where
+    injective prf = injective {f = (g . f)} $ cong g prf
```

Comment:
These should probably be named instances because if you take `id` as one of the functions, they will always be candidates.

Also, `inj` should be kept.

Reply:
Added back `inj` and named the implementations.

Reply:
As far as I can tell it isn't possible to implement `Injective id` because these interfaces can only be implemented for type constructors. Is that intended or is it a bug? Or is there a way to do it?

Reply:
> As far as I can tell it isn't possible to implement `Injective id`

Mmmm. This typechecks fine:

```idris
import Control.Function
                       
Injective Prelude.id where
  injective = id
```

Reply:
Ah, `Prelude.id` works, but `id` doesn't.

Reply:
`id` is a lower-case identifier. Being unqualified, it is quantified and expectedly does not work. It is the same as `Injective a`.